### PR TITLE
convert SDF information to KDL

### PR DESCRIPTION
**Pull request by gustavoneves12**

### CODE REVIEW

File: src/kdl_parser.cpp [hunk 1]
```diff
@@ -171,5 +170,247 @@ bool treeFromUrdfModel(const urdf::ModelInterface& robot_model, Tree& tree)
...
+        }
+    }
+
+    return KDL::RigidBodyInertia();
+}
+
+/**
+ * convert <joint> element to KDL::Joint
+ */
+KDL::Joint sdf_joint_to_kdl(sdf::ElementPtr sdf){
```
Comment (doudou) [inline src/kdl_parser.cpp:246]: wrong naming convention ... you (and the rest of the code) uses lowerCamelCase everywhere else.

File: src/kdl_parser.cpp [hunk 2]
```diff
@@ -171,5 +170,247 @@ bool treeFromUrdfModel(const urdf::ModelInterface& robot_model, Tree& tree)
...
+ * convert <pose> element to KDL::Frame
+ */
+KDL::Frame sdf_pose_to_kdl(sdf::ElementPtr sdf)
+{
+
+    double x, y, z;
+    double roll, pitch, yaw;
+    std::string pose_value = sdf->Get<std::string>();
+
+    sscanf(pose_value.c_str(), "%lf %lf %lf %lf %lf %lf", &x, &y, &z, &roll, &pitch, &yaw);
```
Comment (doudou) [inline src/kdl_parser.cpp]: *validate* sscanf's return value

File: src/kdl_parser.cpp [hunk 3]
```diff
@@ -171,5 +170,247 @@ bool treeFromUrdfModel(const urdf::ModelInterface& robot_model, Tree& tree)
...
+    return KDL::Frame(rotation, position);
+}
+
+/**
+ * convert <inertial> element to KDL::RigidBodyInertia
+ */
+KDL::RigidBodyInertia sdf_inertial_to_kdl(sdf::ElementPtr sdf)
+{
+
+    KDL::Frame pose = sdf_pose_to_kdl(sdf->GetElement("pose"));
```
Comment (doudou) [inline src/kdl_parser.cpp:217]: AFAIK, most pose tags are optional (in which case the pose is the identity)

**Comment by doudou** [comment]
Two minor comments and one major problem:
- move the SDF code into its own file, there's no point keeping everything in one file
- modify treeFromString and treeFromFile to get a third argument of type

```
enum ROBOT_MODEL_FORMAT
{
   ROBOT_MODEL_SDF,
   ROBOT_MODEL_URDF
};
```

and delegate to the right method(s) depending on the argument. The default for this argument should obviously be ROBOT_MODEL_URDF to keep backward compatibility.

**And now for the major issue**

The parsing of poses in SDF is WAAAAY more complicated than that ... I learned that myself when writing the ruby SDF parser.

Basically, all links poses are given relative to the base model. The joint poses are given relative to the joint's child link (unless the &lt;use_parent_model_frame&gt;1&lt;/use_parent_model_frame&gt; tag is given)

AFAIK, to convert to KDL conventions, you will have to compute the joints and links relative poses. By the way, I've used one of the gazebo built-in models (the ur10 model) to test whether my conversion and gazebo's where doing the same thing.

 See https://github.com/Brazilian-Institute-of-Robotics/drivers-transformer/blob/master/ruby/lib/transformer/sdf.rb for the code in Ruby.

File: src/kdl_parser.cpp [hunk 4]
```diff
@@ -171,5 +170,247 @@ bool treeFromUrdfModel(const urdf::ModelInterface& robot_model, Tree& tree)
...
+typedef std::map<std::string, sdf::ElementPtr> LinksMap;
+typedef std::map<std::string, std::vector<std::string> > LinkNamesMap;
+
+/**
+ * convert <axis> element to KDL::Vector
+ */
+KDL::Vector sdf_axis_to_kdl(sdf::ElementPtr sdf)
+{
+    double x, y, z;
+    std::string axis_value = sdf->GetElement("xyz")->Get<std::string>();
```
Comment (doudou) [inline src/kdl_parser.cpp:188]: There is a Vector type in SDF. Would not be surprised that you can do Get&lt;SDF::Vector&gt;

File: src/kdl_parser.cpp (hunk 2 above)
Comment (doudou) [inline src/kdl_parser.cpp:203]: Same comment with SDF::Pose

**Comment by gustavoneves12** [comment]
I talked with Jan about this.

Below you have a text that I sent to Jan.

---

I performed some analyses and found out:
- The link position in the SDF is relative to the model;
- The joint position in the SDF is relative to the child link;
- In gazebo the value of joint position can be updated setting one of three
  types: force, position or velocity.
- The URDF position uses the joint position to set link position.

In my point of view, we don't need go back to using the URDF files. To me
it is not a good idea. We started creating an API to manage SDF files. We
have already shown the visual information from SDF in vizkit3d using the
robot_model library. We only need adjustments.

I thought about that problem and I have two solutions:
1) build the visual information exactly as gazebo does and focus only on
integrating gazebo with vizkit3d. In gazebo we set the joint position using
relative positions. Therefore it is possible build a scene using the SDF,
and afterwards use relative positions of the joints. In my point of view,
it is not a problem.
2) build the visual information using SDF, however, we ignore the pose in
link and set the link position using pose defined in the joint. To me it is
not good.
I would like to do the first solution, because in this case, we are using
the same structure as gazebo.

## 

To solve this I set the segment position to zero. In robot_model I created
some stuff to solve this.

I will commit the robot_model changes and afterwards I will implement your
suggestion.

**Comment by gustavoneves12** [comment]
I am closing PR because I deleted the repository from BIR github account. I was thinking that I was in my github personal account. 

I did fork again and commit, push and created PR again. 

Sorry for my mistake.
